chain/store/newName.go

```diff
@@ -0,0 +1,261 @@
+package store
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/nknorg/nkn/common/serialization"
+)
+
+type nameInfo struct {
+	registrant []byte
+	expiresAt  uint32
+}
+
+type namesCleanup map[string]struct{}
+
+func (ni *nameInfo) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, ni.registrant); err != nil {
+		return err
+	}
+	if err := serialization.WriteUint32(w, ni.expiresAt); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ni *nameInfo) Deserialize(r io.Reader) error {
+	var err error
+	ni.registrant, err = serialization.ReadVarBytes(r)
+	if err != nil {
+		return err
+	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ni *nameInfo) Bytes() ([]byte, error) {
+	buff := bytes.NewBuffer(nil)
+	err := ni.Serialize(buff)
+	if err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
+}
+
+func (ni *nameInfo) String() string {
+	return fmt.Sprintf("%064x %d", ni.registrant, ni.expiresAt)
+}
+
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
+
+func getNewNameId(name string) []byte {
+	nameId := sha256.Sum256([]byte(strings.ToLower(name)))
+	return nameId[:hashPrefixLength]
+}
+
+func (sdb *StateDB) getRegistrant(name string) ([]byte, uint32, error) {
+	nameId := getNewNameId(name)
+	ni, err := sdb.getNameInfo(nameId)
+	if err != nil {
+		return nil, 0, errors.New("name registration info not found")
+	}
+
+	return ni.registrant, ni.expiresAt, nil
+}
+
+func (cs *ChainStore) GetRegistrant(name string) ([]byte, uint32, error) {
+	return cs.States.getRegistrant(name)
```

Comment:
Just discussion:
Should we check the `name string` validity on a public API ?

Reply:
I don't think we need that. When storing pk, the public key has signature verification when verifying transactions so it won't be invalid.

Reply:
`GetRegistrant` not only be invoked by sdb store, but also be invoked by RPC.

Reply:
It shouldn't be a problem because if RPC call has an invalid name, db won't find it.
Also it doesn't make sense to check name string for RPC if name string validation is changed. For example, if for block < 1000, name is valid if len(name) > 3; for block > 1000, name is valid if len(name) > 3. At block 500, a name "alice" is registered. Now at block 2000, someone wants to get registrant of "alice". If we do name validation in RPC, we don't know which rule we should use. If we use the latest rule, name "alice" is invalid, but it actually exists in database.

In short, in order to be consistent, we should only do name validation when registering name. For getting name, we just need to check db because db is the only single source of truth.